lib/puppet/provider/exec/pwsh.rb

```diff
@@ -26,16 +26,28 @@
   EOT
 
   def run(command, check = false)
-    write_script(command) do |native_path|
-      # Ideally, we could keep a handle open on the temp file in this
-      # process (to prevent TOCTOU attacks), and execute powershell
-      # with -File <path>. But powershell complains that it can't open
-      # the file for exclusive access. If we close the handle, then an
-      # attacker could modify the file before we invoke powershell. So
-      # we redirect powershell's stdin to read from the file. Current
-      # versions of Windows use per-user temp directories with strong
-      # permissions, but I'd rather not make (poor) assumptions.
-      return super("cmd.exe /c \"\"#{native_path(command(:pwsh))}\" #{legacy_args} -Command - < \"#{native_path}\"\"", check)
+    if Puppet::Util::Platform.windows?
+      write_script(command) do |native_path|
+        # Ideally, we could keep a handle open on the temp file in this
+        # process (to prevent TOCTOU attacks), and execute powershell
+        # with -File <path>. But powershell complains that it can't open
+        # the file for exclusive access. If we close the handle, then an
+        # attacker could modify the file before we invoke powershell. So
+        # we redirect powershell's stdin to read from the file. Current
+        # versions of Windows use per-user temp directories with strong
+        # permissions, but I'd rather not make (poor) assumptions.
+        return super("cmd.exe /c \"\"#{native_path(command(:pwsh))}\" #{legacy_args} -Command - < \"#{native_path}\"\"", check)
+      end
+    else
+      write_script(command) do |native_path|
+        # Ideally, we could keep a handle open on the temp file in this
+        # process (to prevent TOCTOU attacks), and execute powershell
+        # with -File <path>. But powershell complains that it can't open
+        # the file for exclusive access. If we close the handle, then an
+        # attacker could modify the file before we invoke powershell. So
+        # we redirect powershell's stdin to read from the file.
+        return super("sh -c \"#{native_path(command(:pwsh))} #{posix_args} -Command - < #{native_path}\"", check)
```

Comment:
I'm not sure that `sh` is going to be universally correct across all platforms - so this may be something we want to verify in some manual tests first?

Reply:
should I use `/bin/sh` as per https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/exec/shell.rb#L19 or perhaps get rid of the shell call entirely.

Reply:
Yeah, `/bin/sh` seems sensible if we've already vetted that across other platforms and have standardized on that within `exec` already

Reply:
Fixed